### PR TITLE
KAFKA-15047: Roll active segment when it breaches the retention policy

### PR DESCRIPTION
**Pull request by kamalcph**
Roll the active segment and offload it to remote storage once it breaches the retention time policy.

A segment is eligible for deletion once it gets uploaded to the remote storage. We have checks to allow only the passive segments to be uploaded, so the active segment never gets removed at all even if breaches the retention time. For low-throughput/stale topics, the active segment can hold the data beyond the configured retention time by the user. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala [hunk 1]
```diff
@@ -3911,6 +3911,121 @@ class UnifiedLogTest {
...
+        true
+      })
+    assertEquals(10L, log.logSegments.size())
+    assertEquals(log.nonActiveLogSegmentsFrom(0L).asScala.toSeq, deletableSegments.toSeq)
+  }
+
+  @Test
+  def testActiveSegmentDeletionDueToRetentionTimeBreachWithRemoteStorage(): Unit = {
+    val logConfig = LogTestUtils.createLogConfig(indexIntervalBytes = 1, segmentIndexBytes = 12,
+      retentionMs = 3, localRetentionMs = 1, fileDeleteDelayMs = 0, remoteLogStorageEnable = true)
```
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala:4000]: `localRetentionMs` is set to 1 ms, once it passed the active segment will be rolled over and offloaded to remote storage.

For `localRetentionBytes`, the [UnifiedLog#deleteRetentionSizeBreachedSegments](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L1502-1513) skips the active segment from the size calculation so we cannot assert based on it and we don't want to change the _deleteRetentionSizeBreachedSegments_ behaviour.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                        new KeyValueSpec("k2", "v2"), new KeyValueSpec("k3", "v3"))
+                // consume from the beginning of the topic to read data from local and remote storage
+                .expectFetchFromTieredStorage(broker0, topicA, p0, 4)
+                .consume(topicA, p0, 0L, 4, 4);
+    }
+
+    private Map<String, String> configsToBeAdded() {
+        // Update localLog retentionMs to 1 ms and segment roll-time to 10 ms
+        Map<String, String> topicConfigs = new HashMap<>();
+        topicConfigs.put(TopicConfig.LOCAL_LOG_RETENTION_MS_CONFIG, "1");
```
Comment (kamalcph) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java:72]: `localRetentionMs` is set to 1 ms, once it passed, the active segment will be rolled over and offloaded to the remote storage.

Note that all the tests under tiered-storage integration test specifies the `localRetentionBytes` as 1 byte (See: [TieredStorageTestUtils](https://sourcegraph.com/github.com/apache/kafka/-/blob/storage/src/test/java/org/apache/kafka/tiered/storage/utils/TieredStorageTestUtils.java?L174)) to offload the passive segment ASAP and to keep the active segment on local storage to assert reading the data from both local and remote storage. We don't want to change the [deleteRetentionSizeBreachedSegments](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L1502-1513) behaviour.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                .expectSegmentToBeOffloaded(broker0, topicA, p0, 0, new KeyValueSpec("k0", "v0"))
+                .expectSegmentToBeOffloaded(broker0, topicA, p0, 1, new KeyValueSpec("k1", "v1"))
+                .expectSegmentToBeOffloaded(broker0, topicA, p0, 2, new KeyValueSpec("k2", "v2"))
+                .expectSegmentToBeOffloaded(broker0, topicA, p0, 3, new KeyValueSpec("k3", "v3"))
+                .expectEarliestLocalOffsetInLogDirectory(topicA, p0, 4L)
+                .produce(topicA, p0, new KeyValueSpec("k0", "v0"), new KeyValueSpec("k1", "v1"),
+                        new KeyValueSpec("k2", "v2"), new KeyValueSpec("k3", "v3"))
+                // consume from the beginning of the topic to read data from local and remote storage
+                .expectFetchFromTieredStorage(broker0, topicA, p0, 4)
+                .consume(topicA, p0, 0L, 4, 4);
```
Comment (kamalcph) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java:66]: We can write extensive tests by asserting the log-start-offset, local-log-start-offset, and HW/LSO/LEO once we implement the [OffsetSpec](https://sourcegraph.com/github.com/apache/kafka/-/blob/clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java) for earliestLocal offset.

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 4]
```diff
@@ -1424,25 +1425,71 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
+        val nextSegmentOpt = nextOption(segmentsIterator)
+        val isLastSegmentAndEmpty = nextSegmentOpt.isEmpty && segment.size == 0
+        val upperBoundOffset = if (nextSegmentOpt.nonEmpty) nextSegmentOpt.get.baseOffset() else logEndOffset
+        // We don't delete segments with offsets at or beyond the high watermark to ensure that the log start
+        // offset can never exceed it.
+        val predicateResult = highWatermark >= upperBoundOffset && predicate(segment, nextSegmentOpt)
+
+        // Roll the active segment when it breaches the configured retention policy. The rolled segment will be
+        // eligible for deletion and gets removed in the next iteration.
+        if (predicateResult && !isLastSegmentAndEmpty && remoteLogEnabled() && nextSegmentOpt.isEmpty) {
```
Comment (satishd) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: Can we have the below condition which is easy to understand or some thing better?
```
predicateResult && remoteLogEnabled() &&
nextSegmentOpt.isEmpty  && segment.size > 0 // active segment is not empty
```

Another way is to keep the last segment's predicate result and its size and do the check after the while loop.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java (hunk 3 above)
Comment (satishd) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/RollAndOffloadActiveSegmentTest.java:66]: `OffsetSpec` updates are tracked under https://issues.apache.org/jira/browse/KAFKA-15857

Filed https://issues.apache.org/jira/browse/KAFKA-15864 to track the targeted tests.

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR. Overall LGTM.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the review comments. LGTM.

**Comment by clolov** [comment]
I like this approach because it moves in the direction of us configuring how much data we would like to store locally based on size and time. However, in my head this introduces 2 problems.

The first one is that I no longer see a point in segment.bytes and segment.ms (and by extension log.segment.bytes and log.segment.ms) with respect to tiered topics. If a person says "hey, I only want 4GB of data or data from the last 10 minutes around" then why would they ever need to configure how often a segment should be closed? If this is the case shouldn't this be followed by ignoring those two properties for tiered topics?

The second one is that you will be changing the definition of local.retention. Prior to this change it meant that closed segments will be served from local disk for at most this much size or time as long as they have been moved to tiered storage. Now it will mean that anything beyond this size and time will be found only in tiered storage. Isn't this a "public facing change" and thus requiring some announcements?

**Comment by kamalcph** [comment]
> The first one is that I no longer see a point in segment.bytes and segment.ms (and by extension log.segment.bytes and log.segment.ms) with respect to tiered topics. If a person says "hey, I only want 4GB of data or data from the last 10 minutes around" then why would they ever need to configure how often a segment should be closed? If this is the case shouldn't this be followed by ignoring those two properties for tiered topics?

`segment.bytes` will take effect when the topic has continuous inflow of data. If the user configures `segment.bytes` to 1 GB and `segment.ms` to 1 day and the partition has a bytes-in load of 1 MB/sec, then the segment gets filled in ~40 mins and gets rotated to passive. 

`segment.ms` will take effect when the topic has continuous inflow of data. If the user configures `segment.bytes` to 1 GB and `segment.ms` to 1 day and the partition has a bytes-in load of 100 bytes/sec, then the segment won't be filled and gets rotated to passive after 1 day.

In this patch, we are trying to handle the case where the topic has some data in the active segment but doesn't have continuous inflow of data. So, both the `segment.ms` and `segment.bytes` configs are applicable for tiered storage topics.

> The second one is that you will be changing the definition of local.retention. Prior to this change it meant that closed segments will be served from local disk for at most this much size or time as long as they have been moved to tiered storage. Now it will mean that anything beyond this size and time will be found only in tiered storage. Isn't this a "public facing change" and thus requiring some announcements?

This was the original plan and inline with the local-log segments. If a topic was deprecated and won't be having any more payload in future, the user will expect that all the data in that topic will be removed post the retention time, otherwise it can fail to meet certain compliance.

**Comment by satishd** [comment]
@clolov: The queries that you had in your earlier comments were addressed by @kamalcph, please take a look. I plan to merge these changes if you have no further comments. Thanks.

**Review by clolov — APPROVED** [review]
Yup, apologies, I was planning on responding on Friday. Okay, the explanation makes sense to me currently!

**Comment by satishd** [comment]
Thanks @clolov for the update.

**Comment by satishd** [comment]
@kamalcph The Below tests have failed, please take a look. 

```
Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=zk – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
1m 21s
Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
1m 9s
Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
```

**Comment by kamalcph** [comment]
> @kamalcph The Below tests have failed, please take a look.
> 
> ```
> Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=zk – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
> 1m 21s
> Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
> 1m 9s
> Build / JDK 17 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
> ```

@satishd This seems to be a flaky test and failing even before opening this patch. We can take this separately in KAFKA-15772: 

https://ge.apache.org/scans/tests?search.rootProjectNames=kafka&search.startTimeMax=1699813799999&search.startTimeMin=1698604200000&search.timeZoneId=Asia%2FCalcutta&tests.container=org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest&tests.test=testSendOffsetsWithGroupId(String)%5B1%5D

This test is not failing on other PRs and verified that the test succeeds locally.

**Comment by satishd** [comment]
Thanks @kamalcph for pointing out the JIRA to track the existing intermittent tiered storage related test failure.

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 5]
```diff
@@ -1424,25 +1425,70 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
+        val isLastSegmentAndEmpty = nextSegmentOpt.isEmpty && segment.size == 0
+        val upperBoundOffset = if (nextSegmentOpt.nonEmpty) nextSegmentOpt.get.baseOffset() else logEndOffset
+        // We don't delete segments with offsets at or beyond the high watermark to ensure that the log start
+        // offset can never exceed it.
+        val predicateResult = highWatermark >= upperBoundOffset && predicate(segment, nextSegmentOpt)
+
+        // Roll the active segment when it breaches the configured retention policy. The rolled segment will be
+        // eligible for deletion and gets removed in the next iteration.
+        if (predicateResult && remoteLogEnabled() && nextSegmentOpt.isEmpty && segment.size > 0) {
+          shouldRoll = true
```
Comment (junrao) [inline core/src/main/scala/kafka/log/UnifiedLog.scala:1475]: > We have checks to allow only the passive segments to be uploaded, so the active segment never gets removed at all even if breaches the retention time.

Hmm, the above description of the PR doesn't seem correct. Before this PR, we will consider the deletion of the active segment as long as it's not empty. If the active segment is meets the deletion criteria, we will delete it and automatically roll a new empty segment. So, why do we need to explicitly roll here? Also, why do we only do that when remote log is enabled?

**Review by junrao — COMMENTED** [review]
@kamalcph, @satishd : Sorry for chiming in late here. Added a question for you.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 5 above)
Comment (kamalcph) [reply core/src/main/scala/kafka/log/UnifiedLog.scala:1475]: In the normal topic, we will consider all the segments (inc. active) for deletion due to breach by size/time. When the number of deletable segments equals to the number of segments that exists in the local disk, then we `roll` a new segment and delete all the eligible segments in the same iteration. 

[UnifiedLog#deleteSegments](https://sourcegraph.com/github.com/apache/kafka@21227bda61e75e3a8f1401ff94b27e9161cd3f1b/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L1505-L1511)

With remote storage enabled on the topic, once the active segments gets rolled, we have to successfully upload it to the remote storage before considering it as eligible for deletion so we did this in two iterations:

When the `kafka-log-retention` thread runs:

Iteration-1: If the active segment breached the local-retention time/size, then it rolls the segment to passive.
Iteration-2: If the rolled passive segment gets uploaded to remote storage, then it removes that segment from local disk.

The remote-retention-cleaner thread will remove the uploaded segment from the remote storage when it breaches the complete retention-size/ retention-time.

Comment (junrao) [reply core/src/main/scala/kafka/log/UnifiedLog.scala:1475]: @kamalcph : Thanks for the explanation. I understand this change now.

**Pull request merged**